### PR TITLE
:construction: Abstract classes

### PR DESCRIPTION
**Pull request by simonmarchant**
@MichaelClerx could you have a look at this please and let us know if it looks any good? We just made a lot of abstract classes with the methods we thought would be needed.

### CODE REVIEW

**Comment by simonmarchant** [comment]
PS. tests are failing because we removed the existing tests. Should be okay again once we start implementing things.

File: PKPD/data/abstractData.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+class AbstractData(ABC):
+    def __init__(self):
+        self.datafile = None
+
+    def read_file(self):
+        # read data from a csv (or other) file
+        raise NotImplementedError
+
+    def set_datafile(self):
```
Comment (MichaelClerx) [inline PKPD/data/abstractData.py:11]: Would put this in the constructor or as an argument to read_file ?

**Review by MichaelClerx — COMMENTED** [review]
Hiya!
Looks fairly sensible to me, but hard to predict in advance what you'll need exactly?
It might help to write some example code that uses this API and see if it meets your demands, then remove any methods you didn't use (comment them out?) and make any changes that came up

File: PKPD/data/abstractData.py (hunk 1 above)
Comment (Rebecca-Rumney) [reply PKPD/data/abstractData.py:11]: Hey Michael, 

We've based these abstract classes off of what we did last week, trying to build the classes and get used to the different systems. So all the functions in here are what we think we will need from all we've done so far. This is adaptable as well, so if we realise something needs changing when building, we will.

Comment (simonmarchant) [reply PKPD/data/abstractData.py:11]: > Would put this in the constructor or as an argument to read_file ?

What would that look like in an abstract class? Could we just remove the set_datafile method and when we make the concrete data class add something to the __init__ to set self.datafile to a string containing the filepath?

Comment (MichaelClerx) [reply PKPD/data/abstractData.py:11]: Yeah you can define a default __init__ in an abstract class!
(E.g. https://github.com/pints-team/pints/blob/master/pints/_log_pdfs.py#L83-L106 )

**Comment by martinjrobins** [comment]
> It might help to write some example code that uses this API and see if it meets your demands,

Yup, it might also help to write this as a set of tests as well

**Pull request merged**